refactor(core): group similar error messages

twerske opened the pull request:
group together similar error messages as part of error code efforts

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?
Refactor of errors related to NO_PROVIDER and Circular Dependency in order to be able to group them for runtime error codes.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

AndrewKushnir commented on packages/core/src/render3/di.ts:
```diff
@@ -577,7 +578,7 @@ export function getNodeInjectable(
     const factory: NodeInjectorFactory = value;
     if (factory.resolving) {
-      throw new Error(`Circular dep for ${stringifyForError(tData[index])}`);
+      throwCyclicDependencyError(stringifyForError(tData[index]));
```
It looks like there is one more place where we can reuse `throwCyclicDependencyError` method:
https://github.com/angular/angular/blob/822b838fbcb3dfaaec8b8e800dbbdd903fbfc100/packages/core/src/di/r3_injector.ts#L293-L295

So we can refactor the code there to look like this:
```typescript
      const defName = stringify(defType);
      const path = parents.map(stringify);
      throwCyclicDependencyError(defName, path);
```
and update the `throwCyclicDependencyError` function to also accept path as an argument:
```typescript
export function throwCyclicDependencyError(token: string, path?: string[]): never {
  const depPath = path ? `. Dependency path: ${path.join(' > ')} > ${token}.` : '';
  throw new Error(`Circular dependency in DI detected for ${token}${depPath}`);
}
```

AndrewKushnir reviewed (changes requested):
Great refactoring, thanks @twerske 👍

I left a couple comments and it looks like there is one more place in Ivy runtime code where the `throwCyclicDependencyError` function can be reused (to unify one more error message).

AndrewKushnir commented on packages/compiler/src/provider_analyzer.ts:
```diff
@@ -190,7 +190,7 @@ export class ProviderElementContext {
     if (this._seenProviders.get(tokenReference(token)) != null) {
       this.viewContext.errors.push(new ProviderError(
-          `Cannot instantiate cyclic dependency! ${tokenName(token)}`, this._sourceSpan));
+          `Circular dependency in DI detected for ${tokenName(token)}`, this._sourceSpan));
```
I think we should roll that back, it's a part of VE code. That should also cause some rollbacks in _spec.ts files in your PR.

AndrewKushnir commented on packages/compiler/src/provider_analyzer.ts:
```diff
@@ -366,7 +366,7 @@ export class NgModuleProviderAnalyzer {
     if (this._seenProviders.get(tokenReference(token)) != null) {
       this._errors.push(new ProviderError(
-          `Cannot instantiate cyclic dependency! ${tokenName(token)}`,
+          `Circular dependency in DI detected for ${tokenName(token)}`,
```
This is also a part of VE code and we can rollback.

AndrewKushnir commented on aio/content/guide/dependency-injection-navtree.md:
```diff
@@ -196,7 +196,7 @@ which *is* what parent means.
 2. Angular throws a cyclic dependency error if you omit the `@SkipSelf` decorator.
 
-  `Cannot instantiate cyclic dependency! (BethComponent -> Parent -> BethComponent)`
+  `Circular dependency in DI detected for (BethComponent -> Parent -> BethComponent)`
```
```suggestion
  `Circular dependency in DI detected for BethComponent (BethComponent -> Parent -> BethComponent)`
```

AndrewKushnir reviewed (approved):
@twerske thanks for applying additional updates, the changes look great 👍 

I've also added Jeremy as a reviewer to this PR to see if he has any suggestions for the error message text.

jelbourn reviewed (approved):
LGTM, though I would amend the commit message to go into more detail about which error messages are affected by this change

twerske committed:
refactor(core): group provider and circular errors

group together similar error messages as part of error code efforts
ProviderNotFound & NodeInjector grouped into throwProviderNotFoundError
Cyclic dependency errors grouped into throwCyclicDependencyError

AndrewKushnir commented:
[Presubmit](http://test/OCL:338181361:BASE:338181380:1603245782011:b75a22a3).
